prereise/gather/griddata/hifld/data_process/topology.py

```diff
@@ -0,0 +1,119 @@
+        if tmp_dist < min_dist:
+            min_dist, res_v1, res_v2 = tmp_dist, v1, v2
+    return min_dist, res_v1, res_v2
+
+
+def find_adj_state_of_2_conn_comp(cc1, cc2, state_adj, subs):
+    """Find the adjacency states between two given connected components.
+
+    :param int/iterable cc1: substation ids of the first connected component
+    :param int/iterable cc2: substation ids of the second connected component
+    :param dict state_adj: a dictionary defines state neighbors, keys are state
+        abbreviations, values are list of neighbor state abbreviations.
+    :param pandas.DataFrame subs: substation data frame indexed by substation ids,
+        contains a ``STATE`` column that defines the state of each substation.
+    :return: (*tuple*) -- a pair of sets represents states within each connected
+        component that are adjacent to the other one.
+    """
+    cc1_all_state = set(subs.loc[cc1].STATE.unique())
+    cc2_all_state = set(subs.loc[cc2].STATE.unique())
+    cc1_neighbors = cc1_all_state.copy()
+    cc2_neighbors = cc2_all_state.copy()
+    for s in cc1_all_state:
+        cc1_neighbors |= set(state_adj[s])
+    for s in cc2_all_state:
+        cc2_neighbors |= set(state_adj[s])
+    cc1_adj = cc2_neighbors & cc1_all_state
+    cc2_adj = cc1_neighbors & cc2_all_state
+    return cc1_adj, cc2_adj
+
+
+def connect_islands_with_minimum_cost(
+    lines, subs, cost_metric=haversine, state_neighbor=None
+):
+    """Connect a group of islands defined by lines and substations into one big
+    island with minimum cost.
+
+    :param pandas.DataFrame lines: line data frame indexed by ``ID`` with two columns
+        ``SUB_1_ID`` and ``SUB_2_ID`` representing the corresponding substations at the
+        start and end points.
+    :param pandas.DataFrame subs: substation data frame indexed by ``ID`` with two
+        columns ``LATITUDE`` and ``LONGITUDE`` representing the geographical
+        coordination of each entry.
+    :param function cost_metric: a function defines the cost metric to calculate the
+        weight of lines, defaults to :py:func:`haversine`.
+    :param dict state_neighbor: a dictionary defines the adjacency relationship among
+        states, defaults to None and the constant dictionary ``abv_state_neighbor``
+        defined in the ``const`` module is used.
+    :return: (*tuple*) -- a pair of lists, the first one is a list of all potential
+        line candidates among given connected components; the second one is a list
+        of subsequence of the first entry, representing the chosen lines to form a
+        minimum spanning tree. Each entry of both lists is a 3-tuple, index of the
+        first connected component, index of the second connected component,
+        a dictionary with keys containing ``start``, ``end`` and ``weight``, defines
+        the ``from substation ID``, ``to substation ID`` and the weight of the line
+        calculated by the ``cost_metric``.
```

Comment:
Curious what the use case is for returning the full edge list, I would expect the main goal is to get the MST.

Reply:
We had an offline discussion via Slack regarding caching the full "matrix" for potential subsequent usage. I don't know how we gonna use it at the moment though.